README.md
# Overview

To perform an upgrade using this repo (with very specific parameters in terms of source and target host OS versions):

1. First, create a `params` file locally to use (`cp params.example params`)
1. Set a target commit in the created `params` file
1. Create a file called batch with line-separated UUIDs
1. Run `./run_preupdate.sh` to confirm and prepare all devices in `batch` are ready for upgrade
1. If `grep -i "fail\|error" preupdater.log` returns no output, the update can proceed. Otherwise, some supervisor updates may be required.
1. Finally, run `./run.sh` to perform the actual upgrade and reboot into the new host OS

## Useful helpers

### To generate a device list:
```bash
balena devices -a "${APP}" > device-list.mega.log
```

### To create a batch file based on online status + tag:
```bash
awk '/true.*${STARTING_OS}/{print $NF}' device-list.mega.log | awk -F/ '{print $5}' | sort | shuf > mega-batch.log
for i in $(cat mega-batch.log); do balena tags --device $i | grep -q "${TAG}" && echo $i; done >> tags.log 2>/dev/null
sort -u tags.log | shuf -n "${BATCH_SIZE}" > batch
```

### To filter out devices requiring a supervisor upgrade:
```bash
awk '/Fail: Super/{print $1}' preupdater.log > ../fleetops-supervisor-update/batch
```

### To create a batch for supervisor updates:
```bash
awk '/${STARTING_SUPER}/{print}' device-list.mega.log | grep -f batch | awk -F/ '{print $5}' > supervisor-upgrades
```

### To connect to a device once it returns online:
```bash
uuid="${UUID}"; until balena ssh "${uuid}" 2>/dev/null; do sleep 60; done
```

### To monitor the incremental progress of the update:
```bash
watch -n 120 ./monitor.sh ${APP} batch
```


CHANGELOG.md
## 0.1.0 - 2020-01-08

* Readme: add code blocks [Matthew McGinn]
* Clean up helpers, do not re-reboot, add codeowners & monitor script [Matthew McGinn]
* Debugging: add some helpful commands and extra debugging information [Matthew McGinn]
* Docs: improve notes for larger-scale upgrades [Matthew McGinn]
* Docs: add better README.md for posterity [Matthew McGinn]
* Switch to Github storage for the required update files [Gergely Imreh]
* Adjustments based on experience (#4) [Gergely Imreh]
* Restart supervisor to pick up changes in start script, plus some fixes to run script on Mac (#3) [Pablo Carranza Vélez]


monitor.sh
#!/bin/bash

APP=$1
FILE=${2:-batch}

main(){
    out=$(balena devices -a "${APP}" | grep -f "${FILE}")
    echo "Statuses:"
    echo "${out}" | awk '{print $6}' | sort  | uniq -c
    echo "Online:"
    echo "${out}" | awk '{print $7}' | sort  | uniq -c
    echo "Supervisors Online:"
    echo "${out}" | awk '/true/{print $8}' | sort  | uniq -c
    echo "OSes:"
    echo "${out}" | awk '{print $9}' | sort  | uniq -c
    local -i maybe_need_restart
    maybe_need_restart=$(echo "${out}" | grep -c 'configuring.*true')
    if [ "${maybe_need_restart}" -gt 0 ]; then
        echo "Devices may need supervisor restart:"
        echo "${out}" | awk '/configuring.*true/{print "\t"$NF}'
    fi
    local -i online_not_upgraded
    online_not_upgraded=$(echo "${out}" | grep -c 'true.*Resin')
    if [ "${online_not_upgraded}" -gt 0 ]; then
        echo "Online Laggards:"
        echo "${out}" | awk '/true.*Resin/{print "\t"$NF}'
    fi
    echo "${out}" > device-list.current.log
}

main


updater.sh
#!/bin/bash

set -o errexit -o pipefail

# Don't run anything before this source as it sets PATH here
# shellcheck disable=SC1091
source /etc/profile
# load supervisor version: SUPERVISOR_TAG, SUPERVISOR_IMAGE variables
# shellcheck disable=SC1091
source /etc/resin-supervisor/supervisor.conf
# Load
# shellcheck disable=SC1091
source /usr/sbin/resin-vars

# Linking to a specific git commit sha to always be sure what version of the files are pulled
URLBASE="https://raw.githubusercontent.com/balena-io-playground/fleetops-os-update/ca1996e3b99e07a66220a3291817a0e2aa260d45"
TARGET_OS_VERSION="2.32.0+rev1"
TARGET_SUPERVISOR_REPO="balena/armv7hf-supervisor"
TARGET_SUPERVISOR_VERSION="v9.14.0"
TARGET_OS_VERSION_FILENAME=$(echo "balenaos${TARGET_OS_VERSION}-raspberrypi3.tar.xz" | tr + _)
DOWNLOADS=("xzdec.gz" "rdiff.xz" "hostosupdate.sh" "os/${TARGET_OS_VERSION_FILENAME}" "checksums.txt")
MAX_RETRIES=20

setup_logfile() {
    local workdir=$1
    LOGFILE="${workdir}/osupdate.$(date +"%Y%m%d_%H%M%S").log"
    touch "$LOGFILE"
    tail -f "$LOGFILE" &
    # this is global
    tail_pid=$!
    # redirect all logs to the logfile
    exec 1>> "$LOGFILE" 2>&1
}

stop_supervisor() {
    systemctl stop resin-supervisor || true
    docker rm -f resin_supervisor || true
    systemctl stop update-resin-supervisor.timer || true
}

retry_download() {
    local url=$1
    local curl_retval
    local http_status
    echo "Downloading: ${url}"
    http_status=$(curl -C- --retry "${MAX_RETRIES}" --fail -w "%{http_code}" -O "${url}") || curl_retval=$?
    if [ -n "${curl_retval}" ] && [ "${curl_retval}" -ne 22 ]; then
        finish_up "Curl failure while downloading."
    fi
    if [ -n "${http_status}" ]; then
        # Check if download is successful
        if [ "${http_status}" -ne 200 ]; then
            while [ "${http_status}" -ne 416 ]; do
                # If not finished downlaoding, keep trying
                curl_retval=
                sleep 5;
                echo "Retrying download"
                http_status=$(curl -C- --retry "${MAX_RETRIES}" --fail -w "%{http_code}" -O "$url" ) || curl_retval=$?
                if [ -n "${curl_retval}" ] && [ "${curl_retval}" -ne 22 ]; then
                    finish_up "Curl failure while downloading."
                fi
                if [ "${http_status}" -eq 404 ]; then
                    finish_up "Required file not found on server."
                fi
            done
        fi
        echo "Download finished."
    else
        finish_up "Curl download without final status."
    fi
}

update_supervisor() {
    echo "Updating the supervisor"

    sed -e "s|SUPERVISOR_TAG=.*|SUPERVISOR_TAG=${TARGET_SUPERVISOR_VERSION}|" /etc/resin-supervisor/supervisor.conf > /tmp/update-supervisor.conf
    sed -i -e "s|SUPERVISOR_TAG=.*|SUPERVISOR_TAG=${TARGET_SUPERVISOR_VERSION}|" /etc/resin-supervisor/supervisor.conf

    sed -e "s|SUPERVISOR_IMAGE=.*|SUPERVISOR_IMAGE=${TARGET_SUPERVISOR_REPO}|" /etc/resin-supervisor/supervisor.conf > /tmp/update-supervisor.conf
    sed -i -e "s|SUPERVISOR_IMAGE=.*|SUPERVISOR_IMAGE=${TARGET_SUPERVISOR_REPO}|" /etc/resin-supervisor/supervisor.conf

    update_supervisor_in_api
    # Do not restart the supervisor
}

update_supervisor_in_api() {
    CONFIGJSON="${CONFIG_PATH}"
    TAG="${TARGET_SUPERVISOR_VERSION}"
    APIKEY="$(jq -r '.apiKey // .deviceApiKey' "${CONFIGJSON}")"
    DEVICEID="$(jq -r '.deviceId' "${CONFIGJSON}")"
    API_ENDPOINT="$(jq -r '.apiEndpoint' "${CONFIGJSON}")"
    SLUG="$(jq -r '.deviceType' "${CONFIGJSON}")"
    while ! SUPERVISOR_ID=$(curl -s "${API_ENDPOINT}/v4/supervisor_release?\$select=id,image_name&\$filter=((device_type%20eq%20'$SLUG')%20and%20(supervisor_version%20eq%20'$TAG'))&apikey=${APIKEY}" | jq -e -r '.d[0].id'); do
        echo "Retrying..."
        sleep 5
    done
    echo "Extracted supervisor ID: $SUPERVISOR_ID; setting in the API"
    while ! curl -s "${API_ENDPOINT}/v2/device($DEVICEID)?apikey=$APIKEY" -X PATCH -H 'Content-Type: application/json;charset=UTF-8' --data-binary "{\"supervisor_release\": \"$SUPERVISOR_ID\"}" ; do
        echo "Retrying..."
        sleep 5
    done
}

finish_up() {
    local failure=$1
    local exit_code=0
    if [ -n "${failure}" ]; then
        echo "Fail: ${failure}"
        exit_code=1
    else
        echo "Success; rebooting"
        nohup bash -c "sleep 5 ; reboot " > /dev/null 2>&1 &
    fi
    sleep 2
    kill $tail_pid || true
    exit ${exit_code}
}

main() {
    echo pid $$

    if grep -q "${TARGET_OS_VERSION}" /etc/os-release ; then
       echo "OS update already done, nothing else to do."
       finish_up "OS update already done, nothing else to do."
    fi

    workdir="/mnt/data/ops2"
    mkdir -p "${workdir}" && cd "${workdir}"

    # also sets tail_pid
    setup_logfile "${workdir}"

    delta_name="${SUPERVISOR_TAG}-${TARGET_SUPERVISOR_VERSION}.delta"
    delta_url="${URLBASE}/deltas/${delta_name}.xz"
    echo "Delta URL: ${delta_url}"

    # shellcheck disable=SC1083
    if [ "$(curl -s --head -w %{http_code} "${delta_url}" -o /dev/null)" = "200" ]; then
        echo "Delta found!"
        DOWNLOADS+=("deltas/${delta_name}.xz")
    else
        finish_up "Delta NOT found, bailing!"
    fi

    for item in "${DOWNLOADS[@]}" ; do
        retry_download "${URLBASE}/${item}"
    done

    # Prepare executables
    gunzip -f xzdec.gz && chmod +x xzdec
    ./xzdec rdiff.xz > rdiff && chmod +x rdiff
    chmod +x hostosupdate.sh

    # Extract delta
    ./xzdec "${delta_name}.xz" > "${delta_name}"

    current_supervisor="${SUPERVISOR_IMAGE}:${SUPERVISOR_TAG}"
    output_file="${SUPERVISOR_TAG}"

    echo "Docker save of original supervisor started"
    docker save "${current_supervisor}" > "${output_file}" || finish_up "Docker save error"

    # Work in a subdirectory
    rm -rf workdir || true
    mkdir workdir
    cd workdir
    tar -xf "../${output_file}"
    # shellcheck disable=SC2038
    calculated_sha=$(find . -type f -name layer.tar | xargs sha256sum | sort | grep -v 5f70 | awk '{print $2}'|xargs cat | sha256sum | awk '{ print $1}')
    shipped_sha=$(grep "${SUPERVISOR_TAG}$" ../checksums.txt | awk '{ print $1}')
    if [ "$shipped_sha" != "$calculated_sha" ]; then
        finish_up "Integrity check failure. Expected ${shipped_sha} : got ${calculated_sha}"
    else
        echo "Integrity check okay"
    fi
    # Create delta base
    # shellcheck disable=SC2038
    find . -type f -name layer.tar | xargs sha256sum | sort | grep -v 5f70 | awk '{print $2}' | xargs cat > delta-base

    echo "Creating supervisor image from patch"
    ../rdiff patch delta-base "../${delta_name}" "../${TARGET_SUPERVISOR_VERSION}.tar"

    stop_supervisor

    echo "Docker load supervisor image"
    docker load -i "../${TARGET_SUPERVISOR_VERSION}.tar" || finish_up "Docker load error"

    cd "${workdir}"

    echo "Checking OS update image integrity"
    grep "${TARGET_OS_VERSION_FILENAME}" "checksums.txt" | sha256sum -c -- || finish_up "OS update image integrity check failed"
    echo "Extracting host OS update image"
    ./xzdec "${TARGET_OS_VERSION_FILENAME}" > hostos.tar
    echo "Loading host OS update image into docker"
    docker load -i hostos.tar

    echo "Running host OS update"
    ./hostosupdate.sh --hostos-version "${TARGET_OS_VERSION}" --no-reboot || finish_up "OS update failed"
    # Remove update image
    # shellcheck disable=SC2046
    docker rmi $(docker images | grep resin/resinos | awk '{ print $3}') || true

    echo "Update supervisor"
    update_supervisor

    echo "Cleaning up"
    rm -rf "${workdir}/workdir" || true
    find "${workdir}" -type f ! -name "*.log" -exec rm -rf {} \;

    echo "Finished"

    finish_up
}

(
  # Check if already running and bail if yes
  flock -n 99 || (echo "Already running script..."; exit 1)
  main
) 99>/tmp/updater.lock
# Proper exit, required due to the locking subshell
exit $?


deltas/deltagenerator.sh

TARGET=v9.14.0

generate() {
    base=$1
    echo "Generating $base"
    rm -rf "${base}" || true
    mkdir -p "$base"
    (
        cd ${base} || return
        tar -xf ../${base}.tar
        calculated_sha=$(find -type f -name layer.tar | xargs sha256sum | sort | grep -v 5f70 | awk '{print $2}'|xargs cat | sha256sum | awk '{ print $1}')
        find -type f -name layer.tar | xargs sha256sum | sort | grep -v 5f70 | awk '{print $2}' | xargs cat > delta-base
        rdiff signature --block-size=128 delta-base delta-base.signature
        rdiff delta delta-base.signature ../${TARGET}.tar ../${base}-${TARGET}.delta
        xz -T 0 -9 -e -f ../${base}-${TARGET}.delta
        echo "${calculated_sha}  ${base}" >> ../../checksums.txt
    )
    rm -rf "$base"
}

echo "Clear checksum"
rm ../checksums.txt || true
versions=(v4.0.0 v4.0.0_logstream v4.1.1 v4.1.1_logstream v6.1.2 v6.1.2_logstream v6.2.1 v6.2.5 v6.2.5_logstream v6.6.11_logstream)
for v in ${versions[@]}; do
    generate $v
done


preupdater_template.sh
#!/bin/bash

set -o errexit -o pipefail

# Don't run anything before this source as it sets PATH here
# shellcheck disable=SC1091
source /etc/profile
# Load device settings from config.json
# shellcheck disable=SC1091
source /usr/sbin/resin-vars
# Load SUPERVISOR variables
# shellcheck disable=SC1091
source /etc/resin-supervisor/supervisor.conf

setup_logfile() {
    local workdir=$1
    LOGFILE="${workdir}/preupdate.$(date +"%Y%m%d_%H%M%S").log"
    touch "$LOGFILE"
    tail -f "$LOGFILE" &
    # this is global
    tail_pid=$!
    # redirect all logs to the logfile
    exec 1>> "$LOGFILE" 2>&1
}

finish_up() {
    local failure=$1
    local exit_code=0
    if [ -n "${failure}" ]; then
        echo "Fail: ${failure}"
        exit_code=1
    else
        echo "DONE"
    fi
    sleep 2
    kill $tail_pid || true
    exit ${exit_code}
}

device_repin() {
    COMMIT_HASH=$1
    if [ -z "${COMMIT_HASH}" ]; then
        finish_up "No commit hash provided!"
    fi
    BUILD_ID=$(curl --fail --retry 10 -s "${API_ENDPOINT}/v2/build?\$select=id,commit_hash&\$filter=application%20eq%20${APPLICATION_ID}%20and%20commit_hash%20eq%20'$COMMIT_HASH'" -H "Authorization: Bearer ${DEVICE_API_KEY}" | jq '.d[0].id')  || finish_up "Couldn't get build ID."
    echo "setting device ${DEVICE_ID} to commit ${COMMIT_HASH} with buildID = ${BUILD_ID}"
    curl --fail --retry 10 -s -X PATCH "${API_ENDPOINT}/v2/device(${DEVICE_ID})" -H "Authorization: Bearer ${DEVICE_API_KEY}" -H "Content-Type: application/json" --data-binary '{"build":'"${BUILD_ID}"'}' || finish_up "Couldn't set new commit hash."
}

#######################################
# Globals:
#   CONFIG_PATH
# Arguments:
#   var_name: the name of the required entry in config.json
# Returns:
#   var_value: the value of the entry
#######################################
resin_var_manual_load() {
    local var_name=$1
    local var_value

    if [ -f "${CONFIG_PATH}" ]; then
        var_value=$(jq -r ".${var_name}" "${CONFIG_PATH}") || finish_up "Couldn't get ${var_name} from config.json."
        if [ -z "${var_value}" ]; then
            finish_up "Couldn't load ${var_name} key manually."
        fi
    else
        finish_up "Couldn't find the config.json file."
    fi
    echo "${var_value}"
}

#######################################
# Create or update service environment variable values
# Globals:
#   API_ENDPOINT
#   DEVICE_API_KEY
#   DEVICE_ID
#   UUID
# Arguments:
#   env_name: the env var name
#   env_value: the env var value
# Returns:
#   None
#######################################
service_env_create_update() {
    local env_name=$1
    local env_value=$2
    local ENVID

    ENVID=$(curl --fail --retry 10 -s -X GET \
        "${API_ENDPOINT}/v4/device_service_environment_variable?\$filter=(service_install/device%20eq%20${DEVICE_ID}%20and%20name%20eq%20'${env_name}')" \
        -H "Content-Type: application/json" \
        -H "Authorization: Bearer ${DEVICE_API_KEY}"  | jq .d[0].id) || finish_up "Couldn't get environment variable information correctly."

    if [ "${ENVID}" = "null" ]; then
        echo "Service env var doesn't exists, creating."
        SERVICEID=$(curl --fail --retry 10 -s -X GET "${API_ENDPOINT}/v4/service_install?\$filter=device/uuid%20eq%20'${UUID}'&\$expand=installs__service(\$select=service_name)" -H "Content-Type: application/json" -H "Authorization: Bearer ${DEVICE_API_KEY}" | jq -r .d[0].id)

        curl --fail --retry 10 -s -X POST \
        "${API_ENDPOINT}/v4/device_service_environment_variable" \
        -H "Content-Type: application/json" \
        -H "Authorization: Bearer ${DEVICE_API_KEY}" \
        --data '{
            "service_install": "'"${SERVICEID}"'",
            "name": "'"${env_name}"'",
            "value": "'"${env_value}"'"
        }' || finish_up "Couldn't create service env var."
    else
        echo "Service env var already exists, updating."
        curl --fail --retry 10 -s -X PATCH \
        "${API_ENDPOINT}/v4/device_service_environment_variable(${ENVID})" \
        -H "Content-Type: application/json" \
        -H "Authorization: Bearer ${DEVICE_API_KEY}" \
        --data '{
            "value": "'"${env_value}"'"
        }' || finish_up "Couldn't update service env var."
    fi
}


main() {
    workdir="/mnt/data/ops2"
    mkdir -p "${workdir}" && cd "${workdir}"

    # also sets tail_pid
    setup_logfile "${workdir}"

    # Fill in missing global variables, mostly for 2.0.0-rcX OS versions, that have problem with "resin-var"
    if [ -z "${API_ENDPOINT}" ]; then
        API_ENDPOINT=$(resin_var_manual_load "apiEndpoint")
    fi
    if [ -z "${APPLICATION_ID}" ]; then
        APPLICATION_ID=$(resin_var_manual_load "applicationId")
    fi
    if [ -z "${DEVICE_API_KEY}" ]; then
        DEVICE_API_KEY=$(resin_var_manual_load "deviceApiKey")
    fi
    if [ -z "${DEVICE_ID}" ]; then
        DEVICE_ID=$(resin_var_manual_load "deviceId")
    fi
    if [ -z "$UUID" ]; then
        UUID=$(resin_var_manual_load "uuid")
    fi

    if [ "$SUPERVISOR_TAG" != "v6.6.11_logstream" ]; then
        finish_up "Supervisor needs to be updated to v6.6.11_logstream before continuing"
    fi

    local superstartscript="/usr/bin/start-resin-supervisor"
    # shellcheck disable=SC2016
    if grep -q '^(\[ "$SUPERVISOR_IMAGE_ID".*' "${superstartscript}"; then
        # Modify the supervisor start script so it doesn't recreate the supervisor container on restart
        tempfile=$(mktemp -u /tmp/tmp.supervisor.XXXXXXXXX)
        cp "${superstartscript}" "${tempfile}"

        # shellcheck disable=SC2016
        sed -i 's/^(\[ "$SUPERVISOR_IMAGE_ID".*//' "${tempfile}"
        cat << EOF >> "${tempfile}"
if [ "\$SUPERVISOR_IMAGE_ID" = "\$SUPERVISOR_CONTAINER_IMAGE_ID" ]; then
    docker start --attach resin_supervisor
else
    runSupervisor
fi
EOF
        mount --bind "${tempfile}" "${superstartscript}"
        echo "Supervisor start script was modified."
    else
        echo "Supervisor script is already modified."
    fi

    echo "Restarting supervisor to make sure changes to the start script are picked up"
    systemctl restart resin-supervisor || finish_up "Supervisor restart didn't work."
    sleep 5 # Let the restart commence
    local i=0
    while ! docker ps | grep -q resin_supervisor ; do
      sleep 1
      i=$((i+1))
      if [ $i -gt 60 ] ; then
        finish_up "Supervisor container didn't come up before timeout"
      fi
    done

    echo "Modifying supervisor container"
    docker exec resin_supervisor sed -i 's/waitAsync()\.timeout(5e3)/waitAsync()\.timeout(5e5)/' /usr/src/app/dist/app.js || finish_up "Supervisor container hotfix didn't work."
    echo "Restarting supervisor container"
    systemctl restart resin-supervisor || finish_up "Supervisor restart didn't work."

    echo "Updating environment variable"
    service_env_create_update "DBUS_SYSTEM_BUS_ADDRESS" "unix:path=/run/dbus/system_bus_socket"

    device_repin "%%TARGET_COMMIT%%"

    echo "Finished"

    finish_up
}

(
  # Check if already running and bail if yes
  flock -n 99 || (echo "Already running script..."; exit 1)
  main
) 99>/tmp/updater.lock
# Proper exit, required due to the locking subshell
exit $?
